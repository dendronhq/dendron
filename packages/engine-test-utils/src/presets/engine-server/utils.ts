import { NoteProps, SchemaUtils } from "@dendronhq/common-all";
import {
  CreateNoteFactory,
  NOTE_PRESETS_V4,
  NoteTestUtilsV4,
  PreSetupHookFunction,
  SCHEMA_PRESETS_V4,
} from "@dendronhq/common-test-utils";
import _ from "lodash";
import { resolvePath } from "@dendronhq/common-server";
import path from "path";
import * as fs from "fs";

/**
 * Notes created:
 *   - vault1:
 *     - foo
 *     - foo.ch1
 *   - vault2:
 *     - bar
 */
export const setupBasic: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
  extra,
}) => {
  const vault = vaults[0];
  // TODO: HACK
  let props;
  if (extra?.idv2) {
    props = { id: "foo-id" };
  }
  await NOTE_PRESETS_V4.NOTE_SIMPLE.create({
    vault,
    wsRoot,
    props,
  });
  await NOTE_PRESETS_V4.NOTE_SIMPLE_CHILD.create({
    vault,
    wsRoot,
  });
  await NOTE_PRESETS_V4.NOTE_SIMPLE_OTHER.create({
    vault,
    wsRoot,
  });
  await SCHEMA_PRESETS_V4.SCHEMA_SIMPLE.create({ vault, wsRoot });
};

/**
 <pre>
 /vault1
 ├── bar.ch1.gch1.ggch1.md
 ├── bar.ch1.gch1.md
 ├── bar.ch1.md
 ├── bar.md
 ├── foo.ch1.gch1.ggch1.md
 ├── foo.ch1.gch1.md
 ├── foo.ch1.gch2.md
 ├── foo.ch1.md
 ├── foo.ch2.md
 ├── foo.md
 ├── root.md
 └── root.schema.yml

/vault2
 ├── root.md
 └── root.schema.yml

 /vault3
 ├── root.md
 └── root.schema.yml
</pre>
 * */
export const setupHierarchyForLookupTests: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  const opts = {
    vault: vaults[0],
    wsRoot,
  };
  const fnames = [
    "foo",
    "foo.ch1",
    "foo.ch1.gch1",
    "foo.ch1.gch1.ggch1",
    "foo.ch1.gch2",
    "foo.ch2",
    "bar",
    "bar.ch1",
    "bar.ch1.gch1",
    "bar.ch1.gch1.ggch1",
    "goo.ends-with-ch1.no-ch1-by-itself",
  ];

  for (const fname of fnames) {
    await CreateNoteFactory({ fname, body: `${fname} body` }).create(opts);
  }
};

export const setupJournals: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  const vault = vaults[0];
  const names = [
    "daily",
    "daily.journal",
    "daily.journal.2020",
    "daily.journal.2020.07",
    "daily.journal.2020.07.01.one",
    "daily.journal.2020.07.05.two",
  ];
  return Promise.all(
    names.map((fname) => {
      return NoteTestUtilsV4.createNote({
        wsRoot,
        vault,
        fname,
      });
    })
  );
};

// Workspace will look like:
// .
// ├── dendron.code-workspace
// ├── dendron.yml
// ├── vault1
// │   ├── foo.ch1.md
// │   ├── foo.md
// │   ├── foo.schema.yml
// │   ├── root.md
// │   └── root.schema.yml
// ├── vault2
// │   ├── bar.md
// │   ├── root.md
// │   └── root.schema.yml
// └── vault3
//     ├── root.md
//     └── root.schema.yml
export const setupBasicMulti: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  const vault1 = _.find(vaults, { fsPath: "vault1" })!;
  const vault2 = _.find(vaults, { fsPath: "vault2" })!;
  await NOTE_PRESETS_V4.NOTE_SIMPLE.create({
    vault: vault1,
    wsRoot,
  });
  await NOTE_PRESETS_V4.NOTE_SIMPLE_CHILD.create({
    vault: vault1,
    wsRoot,
  });
  await NOTE_PRESETS_V4.NOTE_SIMPLE_OTHER.create({
    vault: vault2,
    wsRoot,
  });
  await SCHEMA_PRESETS_V4.SCHEMA_SIMPLE.create({ vault: vault1, wsRoot });
};

/**
 *
 * - foo
 * ```
 * ![[foo.one]]
 * ```
 *
 * - foo.one
 * ```
 * # Foo.One
 * ![[foo.two]]
 * Regular wikilink: [[foo.two]]
 * ```
 *
 * - foo.two
 * ```
 * # Foo.Two
 * blah
 * ```
 *
 *
 */
export const setupNoteRefRecursive: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
  extra,
}) => {
  const vault = vaults[0];
  // TODO: HACK
  let props;
  if (extra?.idv2) {
    props = { id: "foo-id" };
  }
  await NOTE_PRESETS_V4.NOTE_SIMPLE.create({
    vault,
    wsRoot,
    body: "![[foo.one]]",
    props,
  });
  await NoteTestUtilsV4.createNote({
    vault,
    fname: "foo.one",
    body: ["# Foo.One", `![[foo.two]]`, `Regular wikilink: [[foo.two]]`].join(
      "\n"
    ),
    wsRoot,
    props: {
      id: "foo.one-id",
    },
  });
  await NoteTestUtilsV4.createNote({
    vault,
    fname: "foo.two",
    body: ["# Foo.Two", `blah`].join("\n"),
    wsRoot,
  });
};

/**
 * Diamond schema is laid out such that:
 *    bar
 *   /   \
 * ch1   ch2
 *   \   /
 *    gch
 * */
export const setupSchemaWithDiamondGrandchildren: PreSetupHookFunction = async (
  opts
) => {
  await setupBasic(opts);
  const { wsRoot, vaults } = opts;
  const vault = vaults[0];
  NoteTestUtilsV4.createSchema({
    fname: "bar",
    wsRoot,
    vault,
    modifier: (schema) => {
      const schemas = [
        SchemaUtils.createFromSchemaOpts({
          id: "bar",
          parent: "root",
          fname: "bar",
          children: ["ch1", "ch2"],
          vault,
        }),
        SchemaUtils.createFromSchemaRaw({
          id: "ch1",
          children: ["gch"],
          vault,
        }),
        SchemaUtils.createFromSchemaRaw({
          id: "ch2",
          children: ["gch"],
          vault,
        }),
        SchemaUtils.createFromSchemaRaw({
          id: "gch",
          template: { id: "template.gch", type: "note" },
          vault,
        }),
      ];
      schemas.map((s) => {
        schema.schemas[s.id] = s;
      });
      return schema;
    },
  });
  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "gch template",
    fname: "template.gch",
    vault,
  });
};

export const setupSchemaPreseet: PreSetupHookFunction = async (opts) => {
  await setupBasic(opts);
  const { wsRoot, vaults } = opts;
  const vault = vaults[0];
  NoteTestUtilsV4.createSchema({
    fname: "bar",
    wsRoot,
    vault,
    modifier: (schema) => {
      const schemas = [
        SchemaUtils.createFromSchemaOpts({
          id: "bar",
          parent: "root",
          fname: "bar",
          children: ["ch1", "ch2"],
          vault,
        }),
        SchemaUtils.createFromSchemaRaw({
          id: "ch1",
          template: { id: "bar.template.ch1", type: "note" },
          vault,
        }),
        SchemaUtils.createFromSchemaRaw({
          id: "ch2",
          template: { id: "bar.template.ch2", type: "note" },
          namespace: true,
          vault,
        }),
      ];
      schemas.map((s) => {
        schema.schemas[s.id] = s;
      });
      return schema;
    },
  });
  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "ch1 template",
    fname: "bar.template.ch1",
    props: { tags: "tag-foo" },
    vault,
  });
  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "ch2 template",
    fname: "bar.template.ch2",
    vault,
  });
};

/**
 * Diamond schema is laid out such that:
 *    bar
 *   /   \
 * ch1   ch2 (namespace: true)
 *   \   /
 *    gch
 * */
export const setupSchemaWithDiamondAndParentNamespace: PreSetupHookFunction =
  async (opts) => {
    await setupBasic(opts);
    const { wsRoot, vaults } = opts;
    const vault1 = vaults[0];

    const withDiamond = path.join(
      resolvePath(vault1.fsPath, wsRoot),
      "withDiamond.schema.yml"
    );
    fs.writeFileSync(
      withDiamond,
      `
version: 1
schemas:
  - id: withDiamond
    children:
      - ch1
      - ch2
    title: withDiamond
    parent: root
  - id: ch1
    children:
      - gch
  - id: ch2
    namespace: true
    children:
      - gch
  - id: gch
    template: template.test
`
    );

    await NoteTestUtilsV4.createNote({
      wsRoot,
      body: "Template text",
      fname: "template.test",
      vault: vault1,
    });
  };

/**
 * Sets up schema which includes a schema that has Diamond grandchildren
 *
 * */
export const setupSchemaWithIncludeOfDiamond: PreSetupHookFunction = async (
  opts
) => {
  await setupBasic(opts);
  const { wsRoot, vaults } = opts;
  const vault1 = vaults[0];

  const withDiamond = path.join(
    resolvePath(vault1.fsPath, wsRoot),
    "withDiamond.schema.yml"
  );
  fs.writeFileSync(
    withDiamond,
    `
version: 1
schemas:
  - id: withDiamond
    children:
      - ch1
      - ch2
    title: withDiamond
    parent: root
  - id: ch1
    children:
      - gch
  - id: ch2
    children:
      - gch
  - id: gch
    template: template.test
`
  );

  const includesDiamondPath = path.join(
    resolvePath(vault1.fsPath, wsRoot),
    "includesDiamond.schema.yml"
  );
  fs.writeFileSync(
    includesDiamondPath,
    `
version: 1

imports:
  - withDiamond

schemas:
  - id: includesDiamond
    parent: root
    namespace: true
    children:
      - a-ch1
      - a-ch2
  - id: a-ch1
    children:
      - withDiamond.gch
  - id: a-ch2
    children:
      - withDiamond.gch
`
  );

  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "Template text",
    fname: "template.test",
    vault: vault1,
  });
};

/**
 * Sets up workspace which has a schema that uses YAML expansion syntax ('<<' type of expansion). */
export const setupSchemaWithExpansion: PreSetupHookFunction = async (opts) => {
  await setupBasic(opts);
  const { wsRoot, vaults } = opts;
  const vault1 = vaults[0];

  const withExpansion = path.join(
    resolvePath(vault1.fsPath, wsRoot),
    "withExpansion.schema.yml"
  );
  fs.writeFileSync(
    withExpansion,
    `
_anchors:
  projects: &projects
    title: projects
    parent: root
    template: templates.projects
    
version: 1
imports: []

schemas:
  - <<: *projects
    id: proj
  - <<: *projects
    id: op
`
  );

  const inlineSchemaPath = path.join(
    resolvePath(vault1.fsPath, wsRoot),
    "includesExpansion.schema.yml"
  );
  fs.writeFileSync(
    inlineSchemaPath,
    `
version: 1

imports:
  - withExpansion

schemas:
  - id: includer
    parent: root
    namespace: true
    children:
      - withExpansion.proj
`
  );

  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "Template text",
    fname: "templates.projects",
    vault: vault1,
  });
};

export const setupInlineSchema: PreSetupHookFunction = async (opts) => {
  await setupBasic(opts);
  const { wsRoot, vaults } = opts;
  const vault1 = vaults[0];

  const inlineSchemaPath = path.join(
    resolvePath(vault1.fsPath, wsRoot),
    "inlined.schema.yml"
  );
  fs.writeFileSync(
    inlineSchemaPath,
    `
version: 1
imports: 
  - foo
schemas:
  - id: plain_schema
    parent: root
    children:
      - plain_schema_child
      - daily
  - id: daily
    parent: root
    children:
      - id: journal
        children:
          - pattern: "[0-2][0-9][0-9][0-9]"
            title: year
            id: year_id
            children:
              - pattern: "[0-1][0-9]"
                children:
                  - pattern: "[0-3][0-9]"
                    title: day
                    template:
                      id: templates.day
                      type: note
  - id: id_with_imported_child
    children:
      - foo.foo
  - id: with_child_that_has_untyped_template
    children:
      - pattern: has_untyped_template
        template: templates.untyped
  - id: plain_schema_child
    template: templates.example
`
  );

  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "Template text",
    fname: "templates.day",
    vault: vault1,
  });

  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "Template text",
    fname: "templates.example",
    vault: vault1,
  });

  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "Untyped template",
    fname: "templates.untyped",
    vault: vault1,
  });
};

export const setupSchemaPresetWithNamespaceTemplateBase: PreSetupHookFunction =
  async (opts) => {
    await setupBasic(opts);
    const { wsRoot, vaults } = opts;
    const vault1 = vaults[0];
    const vault2 = vaults[1];
    NoteTestUtilsV4.createSchema({
      fname: "journal",
      wsRoot,
      vault: vault1,
      modifier: (schema) => {
        const schemas = [
          SchemaUtils.createFromSchemaOpts({
            id: "daily",
            parent: "root",
            fname: "daily",
            children: ["journal"],
            vault: vault1,
          }),
          SchemaUtils.createFromSchemaRaw({
            id: "journal",
            children: ["year"],
            vault: vault1,
          }),
          SchemaUtils.createFromSchemaRaw({
            id: "year",
            pattern: "[0-2][0-9][0-9][0-9]",
            children: ["month"],
            vault: vault1,
          }),
          SchemaUtils.createFromSchemaRaw({
            id: "month",
            pattern: "[0-9][0-9]",
            children: ["day"],
            vault: vault1,
          }),
          SchemaUtils.createFromSchemaRaw({
            id: "day",
            pattern: "[0-9][0-9]",
            namespace: true,
            template: {
              id: "journal.template",
              type: "note",
            },
            vault: vault2,
          }),
        ];
        schemas.map((s) => {
          schema.schemas[s.id] = s;
        });
        return schema;
      },
    });
    await NoteTestUtilsV4.createNote({
      wsRoot,
      body: "Journal",
      fname: "daily",
      vault: vault2,
    });
    await NoteTestUtilsV4.createNote({
      wsRoot,
      body: "Template text",
      fname: "journal.template",
      vault: vault2,
    });
  };

export const setupSchemaPresetWithNamespaceTemplateMulti: PreSetupHookFunction =
  async (opts) => {
    return setupSchemaPresetWithNamespaceTemplateBase(opts);
  };

export const setupSchemaPresetWithNamespaceTemplate: PreSetupHookFunction =
  async (opts) => {
    const vault = opts.vaults[0];
    return setupSchemaPresetWithNamespaceTemplateBase({
      ...opts,
      vaults: [vault, vault],
    });
  };

export const setupEmpty: PreSetupHookFunction = async ({ vaults, wsRoot }) => {
  const vault = vaults[0];
  await NOTE_PRESETS_V4.NOTE_EMPTY.create({
    vault,
    wsRoot,
  });
};

export const setupLinks: PreSetupHookFunction = async ({ vaults, wsRoot }) => {
  return setupLinksBase({ wsRoot, vaults: [vaults[0], vaults[0]] });
};

export const setupLinksMulti: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  return setupLinksBase({ wsRoot, vaults });
};

export const setupLinksBase: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  const vault1 = vaults[0];
  const vault2 = vaults[1];
  // create note with wikilink
  await NOTE_PRESETS_V4.NOTE_WITH_TARGET.create({
    vault: vault1,
    wsRoot,
  });
  // create note with relative wikilink
  await NOTE_PRESETS_V4.NOTE_WITH_ANCHOR_LINK.create({
    vault: vault2,
    wsRoot,
  });
  // create note with labeld wikilink
  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "[[some label|beta]]",
    fname: "omega",
    vault: vault1,
  });
};

/** Creates 2 notes with same fname in 2 different vaults, and a note named
 * "test" in second vault with both valid and invalid wikilinks.
 *
 * See [[scratch.2021.07.15.205433.inconsistent-ref-and-link-behavior]] for the
 * invalid behaviors that this is intended to test for. The only difference is
 * that vaultThree is the real vault and vault3 is the bad one.
 *
 * @returns the test note with the wikilinks.
 */
export const setupMultiVaultSameFname: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}): Promise<NoteProps> => {
  await NoteTestUtilsV4.createNote({
    fname: "eggs",
    vault: vaults[0],
    body: "vault 0",
    wsRoot,
    props: { id: "eggs-vault-0" },
  });
  await NoteTestUtilsV4.createNote({
    fname: "eggs",
    vault: vaults[1],
    body: "vault 1",
    wsRoot,
    props: { id: "eggs-vault-1" },
  });
  return NoteTestUtilsV4.createNote({
    fname: "test",
    vault: vaults[1],
    body: [
      "[[eggs]]", // 7
      "[[dendron://vault1/eggs]]", // 8
      "[[dendron://vault2/eggs]]", // 9
      "[[dendron://vaultThree/eggs]]", // 10
      "[[dendron://vault3/eggs]]", // 11
      "",
      "the test note",
    ].join("\n"),
    wsRoot,
  });
};

export const setupLinksWithVaultBase: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  await setupLinksBase({ vaults, wsRoot });
};

export const setupRefs: PreSetupHookFunction = async ({ vaults, wsRoot }) => {
  const vault = vaults[0];
  // create note with note reference
  await NOTE_PRESETS_V4.NOTE_WITH_NOTE_REF_SIMPLE.create({
    vault,
    wsRoot,
  });
  await NOTE_PRESETS_V4.NOTE_WITH_NOTE_REF_SIMPLE_TARGET.create({
    vault,
    wsRoot,
  });
  // create note with block reference
  await NOTE_PRESETS_V4.NOTE_WITH_BLOCK_REF_SIMPLE.create({
    vault,
    wsRoot,
  });
  // create note with block range reference
  await NOTE_PRESETS_V4.NOTE_WITH_BLOCK_RANGE_REF_SIMPLE.create({
    vault,
    wsRoot,
  });
  // create note with reference offset
  await NOTE_PRESETS_V4.NOTE_WITH_REF_OFFSET.create({
    vault,
    wsRoot,
  });
  // create note with wildcard child reference
  await NOTE_PRESETS_V4.NOTE_WITH_WILDCARD_CHILD_REF.create({
    vault,
    wsRoot,
  });
  // create note with wildcard header reference
  await NOTE_PRESETS_V4.NOTE_WITH_WILDCARD_HEADER_REF.create({
    vault,
    wsRoot,
  });
  // create note with complex wildcard reference
  await NOTE_PRESETS_V4.NOTE_WITH_WILDCARD_COMPLEX.create({
    vault,
    wsRoot,
  });
  // create note with fm variables
  await NOTE_PRESETS_V4.NOTE_WITH_FM_VARIABLES.create({
    vault,
    wsRoot,
  });
};

export const ENGINE_HOOKS_BASE = {
  WITH_LINKS: setupLinksBase,
};

export const ENGINE_HOOKS = {
  setupBasic,
  setupHierarchyForLookupTests,
  setupSchemaPreseet,
  setupSchemaWithDiamondGrandchildren,
  setupSchemaWithIncludeOfDiamond,
  setupSchemaWithDiamondAndParentNamespace,
  setupSchemaPresetWithNamespaceTemplate,
  setupInlineSchema,
  setupSchemaWithExpansion,
  setupNoteRefRecursive,
  setupJournals,
  setupEmpty,
  setupLinks,
  setupRefs,
};

export const ENGINE_HOOKS_MULTI = {
  setupBasicMulti,
  setupLinksMulti,
  setupSchemaPresetWithNamespaceTemplateMulti,
  setupMultiVaultSameFname,
};
