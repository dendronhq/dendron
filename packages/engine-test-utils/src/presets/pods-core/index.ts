import JSON from "./json";
import MARKDOWN from "./markdown";

export const PODS_CORE = {
  JSON,
  MARKDOWN,
};

export const PODS_PRESETS = [{ name: "json", presets: JSON }];
