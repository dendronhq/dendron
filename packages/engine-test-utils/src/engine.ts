import {
  CleanDendronSiteConfig,
  CONSTANTS, DendronConfig, DEngineClient,
  DVault,
  DWorkspace,
  WorkspaceFolderRaw,
  WorkspaceOpts,
  WorkspaceSettings
} from "@dendronhq/common-all";
import {
  getDurationMilliseconds,
  tmpDir,
  vault2Path
} from "@dendronhq/common-server";
import {
  GenTestResults,
  PostSetupHookFunction,
  PreSetupHookFunction,
  RunEngineTestFunctionOpts,
  RunEngineTestFunctionV4,
  runJestHarnessV2,
  SetupHookFunction,
  SetupTestFunctionV4,
  sinon,
  TestResult
} from "@dendronhq/common-test-utils";
import { LaunchEngineServerCommand } from "@dendronhq/dendron-cli";
import {
  createEngine as engineServerCreateEngine,
  DConfig,
  WorkspaceService
} from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import os from "os";
import path from "path";
import { SinonStub } from "sinon";
import { ENGINE_HOOKS } from "./presets";
import { GitTestUtils } from "./utils";

export type TestSetupWorkspaceOpts = {
  /**
   * Vaults to initialize engine with
   * Defaults to following if not set
   * [
   *    { fsPath: "vault1" },
   *    { fsPath: "vault2" },
   *    { fsPath: "vault3", name: "vaultThree" },
   *  ]
   */
  vaults?: DVault[];
  /**
   * Modify dendron config before initialization
   */
  modConfigCb?: (config: DendronConfig) => DendronConfig;
};

export type AsyncCreateEngineFunction = (
  opts: WorkspaceOpts
) => Promise<{engine: DEngineClient, port?: number}>;

/**
 * Create an {@link DendronEngine}
 */
export async function createEngineFromEngine(opts: WorkspaceOpts) {
  return {engine: engineServerCreateEngine(opts) as DEngineClient, port: undefined};
}

export { DEngineClient, DVault, WorkspaceOpts };

/**
 * Create a server
 * @param opts
 * @returns
 */
// @ts-ignore
export async function createServer(opts: WorkspaceOpts) {
  return await new LaunchEngineServerCommand().enrichArgs({
    wsRoot: opts.wsRoot,
  });
}

/**
 * Create an {@link DendronEngineClient}
 */
export async function createEngineFromServer(
  opts: WorkspaceOpts
): Promise<any> {
  const { engine, port } = await createServer(opts);
  await engine.init();
  return {engine, port};
}

export function createSiteConfig(
  opts: Partial<CleanDendronSiteConfig> &
    Required<Pick<CleanDendronSiteConfig, "siteRootDir" | "siteHierarchies">>
): CleanDendronSiteConfig {
  let copts = {
    siteNotesDir: "docs",
    siteUrl: "https://localhost:8080",
    ...opts,
  };
  return {
    ...copts,
    siteIndex: DConfig.getSiteIndex(copts),
  };
}

/**
 *
 * @param opts.asRemote: add git repo
 * @param opts.wsRoot: override given wsRoot
 * @returns
 */
export async function setupWS(opts: {
  vaults: DVault[];
  workspaces?: DWorkspace[];
  asRemote?: boolean;
  wsRoot?: string;
}) {
  const wsRoot = opts.wsRoot || tmpDir().name;
  const ws = new WorkspaceService({ wsRoot });
  ws.createConfig();
  const config = ws.config;
  let vaults = await Promise.all(
    opts.vaults.map(async (vault) => {
      await ws.createVault({ vault, config, updateConfig: false });
      return vault;
    })
  );
  config.vaults = _.sortBy(config.vaults, "fsPath");
  if (config.site.duplicateNoteBehavior) {
    config.site.duplicateNoteBehavior.payload = (
      config.site.duplicateNoteBehavior.payload as string[]
    ).sort();
  }
  ws.setConfig(config);
  if (opts.workspaces) {
    const vaultsFromWs = await _.reduce(
      opts.workspaces,
      async (resp, ent) => {
        await resp;
        await WorkspaceService.createWorkspace({
          wsRoot: path.join(wsRoot, ent.name),
          vaults: ent.vaults,
        });
        return ws.addWorkspace({ workspace: ent });
      },
      Promise.resolve({ vaults: [] } as { vaults: DVault[] })
    );
    vaults = vaults.concat(vaultsFromWs.vaults);
  }
  if (opts.asRemote) {
    await GitTestUtils.createRepoWithReadme(wsRoot);
  }
  return { wsRoot, vaults };
}

export type RunEngineTestV5Opts = {
  preSetupHook?: SetupHookFunction;
  createEngine?: AsyncCreateEngineFunction;
  extra?: any;
  expect: any;
  workspaces?: DWorkspace[];
  setupOnly?: boolean;
  initGit?: boolean;
  initHooks?: boolean;
  addVSWorkspace?: boolean;
  /**
   * Path to preset wsRoot
   */
  wsRoot?: string;
} & TestSetupWorkspaceOpts;

export type RunEngineTestFunctionV5<T = any> = (
  opts: RunEngineTestFunctionOpts & { extra?: any; engineInitDuration: number, port?: number }
) => Promise<TestResult[] | void | T>;

export class TestPresetEntryV5 {
  public preSetupHook: PreSetupHookFunction;
  public postSetupHook: PostSetupHookFunction;
  public testFunc: RunEngineTestFunctionV4;
  public extraOpts: any;
  public setupTest?: SetupTestFunctionV4;
  public genTestResults?: GenTestResults;
  public vaults: DVault[];
  public workspaces: DWorkspace[];

  constructor(
    func: RunEngineTestFunctionV5,
    opts?: {
      preSetupHook?: PreSetupHookFunction;
      postSetupHook?: PostSetupHookFunction;
      extraOpts?: any;
      setupTest?: SetupTestFunctionV4;
      genTestResults?: GenTestResults;
      vaults?: DVault[];
    }
  ) {
    let {
      preSetupHook,
      postSetupHook,
      extraOpts,
      setupTest,
      genTestResults,
      workspaces,
    } = _.defaults(opts, {
      workspaces: [],
    });
    this.preSetupHook = preSetupHook ? preSetupHook : async () => {};
    this.postSetupHook = postSetupHook ? postSetupHook : async () => {};
    this.testFunc = _.bind(func, this);
    this.extraOpts = extraOpts;
    this.setupTest = setupTest;
    this.genTestResults = _.bind(
      genTestResults ? genTestResults : async () => [],
      this
    );
    this.workspaces = workspaces;
    this.vaults = opts?.vaults || [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      {
        name: "vaultThree",
        fsPath: "vault3",
      },
    ];
  }
}

/**
 *
 * To create empty workspace, initilizae with `vaults = []`
 * @param func
 * @param opts.vaults: By default, initiate 3 vaults {vault1, vault2, (vault3, "vaultThree")}
 * @param opts.preSetupHook: By default, initiate empty
 * @param opts.wsRoot: Override the randomly generated test directory for the wsRoot
 * @returns
 */
export async function runEngineTestV5(
  func: RunEngineTestFunctionV5,
  opts: RunEngineTestV5Opts
): Promise<any> {
  const {
    preSetupHook,
    extra,
    vaults: vaultsInit,
    createEngine,
    initGit,
    workspaces,
    addVSWorkspace,
  } = _.defaults(opts, {
    preSetupHook: async ({}) => {},
    postSetupHook: async ({}) => {},
    createEngine: createEngineFromEngine,
    extra: {},
    // third vault has diff name
    vaults: [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      { fsPath: "vault3", name: "vaultThree" },
    ],
    addVSWorkspace: false,
  });
  try {
    // make sure tests don't overwrite local homedir contents
    TestEngineUtils.mockHomeDir();
    const { wsRoot, vaults } = await setupWS({
      vaults: vaultsInit,
      workspaces,
      wsRoot: opts.wsRoot,
    });
    if ((opts.initHooks, vaults)) {
      fs.ensureDirSync(path.join(wsRoot, CONSTANTS.DENDRON_HOOKS_BASE));
    }
    await preSetupHook({ wsRoot, vaults });
    const resp = await createEngine({ wsRoot, vaults });
    const engine = resp.engine;
    const start = process.hrtime();
    const initResp = await engine.init();
    if (addVSWorkspace) {
      fs.writeJSONSync(
        path.join(wsRoot, CONSTANTS.DENDRON_WS_NAME),
        {
          folders: vaults.map((ent) => ({
            path: ent.fsPath,
            name: ent.name,
          })) as WorkspaceFolderRaw[],
          settings: {},
          extensions: {},
        } as WorkspaceSettings,
        { spaces: 4 }
      );
    }
    const engineInitDuration = getDurationMilliseconds(start);
    const testOpts = {
      wsRoot,
      vaults,
      engine,
      initResp,
      port: resp.port,
      extra,
      config: engine,
      engineInitDuration,
    };
    if (initGit) {
      await GitTestUtils.createRepoForWorkspace(wsRoot);
      await Promise.all(
        vaults.map((vault) => {
          return GitTestUtils.createRepoWithReadme(
            vault2Path({ vault, wsRoot })
          );
        })
      );
    }
    if (opts.setupOnly) {
      return testOpts;
    }
    const results = (await func(testOpts)) || [];
    await runJestHarnessV2(results, expect);
    return { opts: testOpts, resp: undefined, wsRoot };
  } finally {
    // restore sinon so other tests can keep running
    sinon.restore();
  }
}

export function testWithEngine(
  prompt: string,
  func: RunEngineTestFunctionV4,
  opts?: Omit<RunEngineTestV5Opts, "expect"> & { only?: boolean }
) {
  if (opts?.only) {
    return test.only(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  } else {
    return test(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  }
}

export class TestEngineUtils {
  static mockHomeDir(dir?: string): SinonStub {
    if (_.isUndefined(dir)) dir = tmpDir().name;
    return sinon.stub(os, "homedir").returns(dir);
  }

  static vault1(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault1" })!;
  }

  static vault2(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault2" })!;
  }

  static vault3(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault3" })!;
  }
}
