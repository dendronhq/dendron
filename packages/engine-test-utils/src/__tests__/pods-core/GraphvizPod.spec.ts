import { tmpDir } from "@dendronhq/common-server";
import { ENGINE_HOOKS, FileTestUtils } from "@dendronhq/common-test-utils";
import { GraphvizExportPod } from "@dendronhq/pods-core";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";
import { runEngineTestV5 } from "../../engine";
import { checkString } from "../../utils";

describe("graphviz export pod", () => {
  let exportDest: string;

  beforeAll(() => {
    exportDest = tmpDir().name;
  });

  test("basic", async () => {
    await runEngineTestV5(
      async ({ engine, vaults, wsRoot }) => {
        const pod = new GraphvizExportPod();
        engine.config.useFMTitle = true;
        await pod.execute({
          engine,
          vaults,
          wsRoot,
          config: {
            dest: exportDest,
            includeBody: true,
            includeStubs: true,
          },
        });

        // check that graphviz file is created
        let [expectedFiles, actualFiles] = FileTestUtils.cmpFiles(exportDest, [
          "graphviz.dot",
        ]);
        expect(expectedFiles).toEqual(actualFiles);

        // check contents of graphviz file
        const foo = fs.readFileSync(path.join(exportDest, "graphviz.dot"), {
          encoding: "utf8",
        });

        // TODO: IDs of files switch between test runs, which causes the snapshot
        // to fail when it shouldn't
        // expect(foo).toMatchSnapshot("graphviz contents");

        // Check for:
        // 1. start of file -> "graph {"
        // 2. existence of a note -> "note_"
        // 3. a labeled note -> "[label=\""
        // 4. a connection -> "--"
        // 5. end of graph -> "}"
        await checkString(foo, "graph {", "note_", '[label="', "--", "}");
      },
      { expect, preSetupHook: ENGINE_HOOKS.setupBasic }
    );
  });
});
