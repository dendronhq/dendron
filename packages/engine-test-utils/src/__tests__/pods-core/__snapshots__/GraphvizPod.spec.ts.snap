// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphviz export pod basic: graphviz contents 1`] = `
"graph {
    note_0a4dc8f3033b4488a99e69e9707ae106 [label=\\"Root\\"];
	note_bar [label=\\"Bar\\"];
	note_0a4dc8f3033b4488a99e69e9707ae106 -- note_bar;
	note_foo [label=\\"Foo\\"];
	note_0a4dc8f3033b4488a99e69e9707ae106 -- note_foo;
	note_foo.ch1 [label=\\"Ch1\\"];
	note_foo -- note_foo.ch1;
	note_0c317cd06a424382bb4c1dc09a39ebd7 [label=\\"Root\\"];
	note_7bc74f7b3de040d09319d120576bb9d4 [label=\\"Root\\"];
}"
`;
